### DISCUSSION TITLE
VRM instance toggles after every reboot

### DISCUSSION THREAD

**Discussion by winfriedp** [Louisvdw/dbus-serialbattery#481] [Q&A] [answered]
dbus-serialbattery 0.14.3 is installed on a Cerbo GX v2.93. My battery is connected from a JK BMS to the Cerbo via an isolated USB to RS485 Converter with a CH340E Chip. Every time I reboot the Cerbo the SerialBattery VRM Instance toggles between 1 and 2. Has anyone ever observed the behavior?

**Comment by teefixx** [comment]
Hi, I see the same behaviour and I think this issue (https://github.com/Louisvdw/dbus-serialbattery/issues/456) is caused by that fact.

**Reply by winfriedp** [reply]
Thanks for the advice. It is quite possible that this is the same root cause, but what is the root cause? SerialBattery, the connection adapter or Venus OS? I'm afraid only Louis can answer that.

**Comment by teefixx** [comment]
Just as additional information: I just had the problem, I had to restart my VenusOS
- got new VRM Instances

Tried another reboot and I got the original VRM instances assigned. I think this is by chance...

**Comment by DirkReudenbach** [comment]
On my Installation it´s changing always between the VRM instance 1-3 after reboot. It‘s even more worth because if connected with Cerbogx and BMS you need to choose ohne BMS as reference. If the instance is changing after reboot - the MPPT will fail with Code67 which results in no charge of the MPPTs. Therefore ím using the Smart Shunt as reference. Maybe Louis can investigate?

**Comment by mr-manuel** [comment, marked as answer]
Have you all only one BMS or multiple BMS?

1. Could you check, if your serial to USB adapter changes also ID like `ttyUSB0` to `ttyUSB1`?
2. Check with `dbus-spy` the value under `com.victronenergy.settings -> Settings/Devices/serialbattery_zero/ClassAndVrmInstance`

This should be the relevant lines to this: https://github.com/Louisvdw/dbus-serialbattery/blob/6471152bf701eb46beaa9ba4d11d93ffd1559dd4/etc/dbus-serialbattery/dbushelper.py#L42-L84

A possible solution would be to bind the `DeviceInstance` to the serial number of the BMS, so it also doesn't matter, if you change adapter.

**Reply by teefixx** [reply]
So far it didn't change but what's odd is, that I have serialbattery with ID 1,2,3 but only two batteries...

<img width="1233" alt="image" src="https://user-images.githubusercontent.com/35899354/223982296-4c620df8-66c3-450b-89ce-dc2fe2dad791.png">

**Reply by mr-manuel** [reply]
Maybe you have 3 serial to USB adapters? The driver at the moment checks all, even if there is no BMS connected.

**Reply by teefixx** [reply]
Indeed I do, there are EM540 on RS485 / USB adapters connected.

**Reply by winfriedp** [reply]
Me too. I have two serial adapters, one for JK BMS and one for a grid-meter. Now I bind the two ports with udev rules that create symlinks. Then I boot my Cerbo serveral times and the ports no longer toggle and remain stable. I hope my problem is now solved. Thanks mr-manuel for the thought.

**Reply by mr-manuel** [reply]
If the adapters have different ID then it works with udev rules. The problem is when someone uses multiple serial to USB adapters that have the same ID.

You can check this with `lsusb`.

**Reply by teefixx** [reply]
That's the case with my installation: I use two RS485/USB adapters for two batteries, they have the same ID... The one for the two EM540 has a different ID.

**Reply by DirkReudenbach** [reply]
I'm not a programmer - so sorry for maybe easy questions.

After checking with mr-manuel advices - in my Installation with 2 JK-BMS and 1 grid meter are 4 USB adapters present.  The ID's of the JK BMS are changing. What are the next steps to get it done? 

![20230311_081847](https://user-images.githubusercontent.com/119795730/224472552-0d1a79d0-ce13-4259-9212-6aaeec1babe5.jpg)

**Reply by mr-manuel** [reply]
At the moment there is no persistent solution, if the adapter has the same ID, but you can try https://github.com/Louisvdw/dbus-serialbattery/issues/456#issuecomment-1461966263

**Reply by winfriedp** [reply]
Here a short instruction of my solution with using udev rules creating a symlink for two usb devices:
Using a udev rule you need an unique selection criteria for the device. In my case the vendor- and product-Id.

Identify vendor- and product-Id:
```
root@einstein:/dev# lsusb
Bus 004 Device 003: ID 1a86:7523 QinHeng Electronics HL-340 USB-Serial adapter
Bus 003 Device 002: ID 0403:6001 Future Technology Devices International, Ltd FT232 Serial (UART) IC
```
In this case vendor ids are 1a86 and 0403 and product ids are 7523 and 6001, in usv terms ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523" for mey JK BMS adapter and ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001" for my EM540 adapter.
An additional selection criteria is KERNEL=="ttyUSB*". Maybe it works without this criterion.
The complete two udev rules defining a symlink are:
```
KERNEL=="ttyUSB*", ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523", SYMLINK+="my_jkbms1"
KERNEL=="ttyUSB*", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="my_em540"
```
Create a new rules-file (e.g.  99-usb-serial.rules in /etc/udev/rules.d/, fill in the two rows with the rules and save the file.

Now reload the rules:
`udevadm control --reload-rules && udevadm trigger`

If every step was OK, list the new symlinks:
```
lrwroot@einstein:/dev# ls -l my_*
lrwxxrwxrwx  1 root     root     7 Mar  9 17:18 my_jkbms1 -> ttyUSB0
rwxrwx       1 root     root     7 Mar  9 17:18 my_em540 -> ttyUSB1
```
As Manuel wrote this only works if you have a unique selction criteria for your adapter.
This YT [https://www.youtube.com/watch?v=BOxWrMNXKpU] was the template for my instruction.

**Reply by DirkReudenbach** [reply]
Thanks @winfriedp -> unfortunately I'm using the same adapter twice - so if i would use the udev rules I need to by another adapter or could this issue be solved by installing the BLE release?

![20230312_083000](https://user-images.githubusercontent.com/119795730/224530932-27b4fe6a-03e8-40ff-8f9e-f32ed4210f8e.jpg)

**Reply by mr-manuel** [reply]
> At the moment there is no persistent solution, if the adapter has the same ID, but you can try [#456 (comment)](https://github.com/Louisvdw/dbus-serialbattery/issues/456#issuecomment-1461966263)

I‘m also looking for a solution to assign the `Device Instance` by serial number of the BMS, but it seems not that easy to get the serial number from the BMS. Help to achieve this would be appreciated.

I have different RS485 adapters, but all have the same ID since they probably have the same chipset. If you buy another adapter you should probably use another chipset, but it has to be compatible with Venus OS. You could also try to connect one adapter over a USB Hub, if you already have one. If it‘s possible you could also use a UART/TTL adapter for one BMS.

**Reply by mr-manuel** [reply]
Yes, you can try to connect one battery over serial USB adapter and one over BLE. The connection order should always be the same.

**Reply by DirkReudenbach** [reply]
Thx Manuel - USB Hub already tried - no changes-> think UART or BLE could be an option. Will try and feed back

**Reply by brunnema** [reply]
Hi Guys,
isn't there the Option to chose one criteria by the Listing like the USB port as well? So I had the same issue in an other project where I need to fix this for 3 identical USB Adapters with all the same Serial/ Vendor ID. Using the dedicatd USB port will allways pin a Symlink to the same USB port, disadvantage is that you shouldn't change the Adapter ports on the USB.
BR Martin

**Reply by LarsAndersenDenmark** [reply]
If using a raspberry, does it matter and improve chances of getting the same vrm instances if you instead of an external usb hub, use an extension usb modul for the raspberry?

For now I have had sufficient ports in my raspberry, but need now an extra for connecting the shunt.

**Reply by mr-manuel** [reply]
@brunnema could you post here how you did that?

**Reply by brunnema** [reply]
Hi,
here are brief intros - how to determine a USB device and set it to a dedicated tty connect:

- use these commands for e.g. to determine your USB stuff when connected
#dmesg | grep usb
#lsusb 

- overview which devices are assigened: ls /dev
- use uadev for the wanted device to get fixed realtion with a symlink lateron 
e.g. #udevadm info -a -n /dev/ttyUSB0

-  determine a unique attribute which you can use like Serial, Vendor ID or at least the pluged in USB port (in case there is no unique identify of the device itself)
e.g. you can quick grap the attr if you know them like this
 #udevadm info -a -n /dev/ttyUSB0 | grep '{serial}' | head -n1
#udevadm info -a -n /dev/ttyUSB0 | grep '{serial}' | head -n1
or path: 
#udevadm info -a -n /dev/ttyUSB0 | grep '{devpath}' | head -n1

- to dertine this symlink - a seriall file needs to be added or created:
sudo mcedit /etc/udev/rules.d/99-usb-serial.rules

content should be like this: 
SUBSYSTEM=="tty", ATTRS{serial}=="SERIENNUMMER", SYMLINK+="YOURNAME", OWNER="YourUsedOwner"
e.g. 
// FTDI-Daly BMS auf USB Steckplatz oben links
SUBSYSTEM=="tty", ATTRS{idVendor}=="1a86", SYMLINK+="DALY-BMS", OWNER="pi"
save file and reboot

- then you can check if the symlink is available and reference to it instead of using ttyusb0
ls -la /dev/serial/by-id/
ls -l /dev/

![image](https://user-images.githubusercontent.com/61138742/228909686-e06b4cfc-b401-4bf5-9a0c-d146283b5ecd.png)

Hope this help you creating unique and persisting USB relationship
Best Regards Martin

**Reply by Besenreiter** [reply]
Hello,
my question is because of lack of knowlege about the python code. When I have this unique symlink, where do I put it?? What file do I have to change in what way?
Thanks

**Comment by brunnema** [comment]
Hi all, 
I have a similar questions to avoid thos USB struggeling issue.
Is it possible to run a RS485HAT on the RPI, like the Waveshare 2ch RS485 HAT?
This Board would provide the both RS485 channels as ttysc0 and ttysc1 instead of ttyusbxx
Where can I change the pointer from ttyusb01 to ttysc0 for example to get the BMS run onto a HAT Board.
BR Martin

**Reply by mr-manuel** [reply]
This works out of the box, since it checks all tty*.

https://github.com/Louisvdw/dbus-serialbattery/blob/6471152bf701eb46beaa9ba4d11d93ffd1559dd4/etc/dbus-serialbattery/service/run#L1-L3

**Reply by brunnema** [reply]
Thanks Manuel, good news! 
Then let me check how to get the board initatet and run. BR Martin

**Reply by DirkReudenbach** [reply]
@mr-manuel 
can I use this for my issue with 2 serial batteries? How do I start this sript?

**Reply by mr-manuel** [reply]
1. Do you mean with "this" the RS485HAT?
2. You don't need to start the script. Only install the driver as you would normally do.

**Comment by mdkeil** [comment]
Hello,

how it works with VenusOS 3.x / RPI2 ..I tryed to make some udev-rules for creating new symlinks but "serial battery" further use /dev/ttyUSBX (created by serial starter udev rule?) which are using for setting the VRM Instance ID.. If I set the symlink directly to /dev/ttyUSB0, it will be ignored or overwritten (by serial starter udev rule?)

**Comment by Micha854** [comment]
@brunnema 

Thanks for the tutorial but when I run dbus-spy the devices are still connected via ttyUSB*

don't you have to adjust something on venus here?

example on dbus-spy

com.victronenergy.battery.DALY-BMS

so we would have a really fixed assignment

**Reply by brunnema** [reply]
Hi Micha,
I still need to swap USB-RS485 towards the Waveshare RS485 HAT. 
Sorry, there is a bit of mixture...the Tutorial is only requied in case you want to use multiple USB Sticks/ports to keep them stable linked to the attached Devices (case of reboot, etc).
 A syslink is not requieed, cause dbus script scanns all tty* as posted by Manuel.
In Case I want to swap towards waveshare Hat RS485 it will become ttysc0 and ttysc1 available which are dedicated to each if the 2 channles and there is no port fix requierd as it is on USB ports.
BR Martin
